synthesizer/src/vm/finalize.rs

```diff
@@ -852,10 +854,24 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
         // Initialize the list of deployment payers.
         let mut deployment_payers: IndexSet<Address<N>> = Default::default();
 
-        // Abort the transactions that are have duplicates or are invalid. This will prevent the VM from performing
-        // verification on transactions that would have been aborted in `VM::atomic_speculate`.
+        // Abort fee transactions and invalid deployments. This will prevent the
+        // VM from performing expensive verification on invalid fees or
+        // deployments that would have been aborted in `VM::atomic_speculate`.
         for transaction in transactions.iter() {
-            // Determine if the transaction should be aborted.
+            // If the transaction is an Execution, we will fully verify it.
+            // Executions require full verification to avoid malleability attacks.
+            if transaction.is_execute() {
+                executions.push(*transaction);
+                continue;
+            }
+
+            // Abort the transaction if it is a fee transaction.
+            if transaction.is_fee() {
+                aborted_transactions.push((*transaction, "Fee transactions are not allowed in speculate".to_string()));
+                continue;
+            }
```

Comment:
I believe adding this check should be enough. The only malleability attack validators can do is to convert the Execution to a Fee transaction. 

The `FakeTransaction` attack can only be done this way. Having actual `input_id`, `output_id`, and `tcm` duplicates can only be crafted with the same private key, so malicious validators can't simply inject a transition from another transaction into their own. 

@ghostant-1017 Let me know if this makes sense and if I am missing anything.

Reply:
My understanding is that the attack entails a validator creating a completely invalid bogus transaction by adding a random transition. 

I pushed a proof of concept to the tests: 8949c47d0 . At least at this level in the code, I confirmed that without this PR, the bogus transaction "pushes out" the honest transaction.

Sidenote: This also made me realise that malicious validators can still cause payment-free verification of their malformed executions, though these should be cheap. I'll think about a resolution.

Reply:
I stand corrected. It looks like the validators can also mutate the transactions by swapping out the fee transaction. If the fee is too low, the transaction will be aborted (if the fee is high enough, then the transaction will be rejected and the fee is consumed.


The validator might also be able to mutate the transaction fields with garbage s.t. verification will fail and the TX will be aborted.

Reply:
> The validator might also be able to mutate the transaction fields with garbage s.t. verification will fail and the TX will be aborted.

Good point, I think I also see an angle where the attacker can force a deployment to be aborted. The attacker mutates the deployment's only transition, which is the fee, to change its outputs but keep its inputs. Or vice versa. Therefore the transaction_id will be different but it still partially collides in the `should_abort_transaction` check. 

Despite this, I have a weak opinion to keep calling `should_abort_transaction` on deployments, as this PR does, as it is still an effective DoS protection measure.

In general, we can consider proper malleability protection measures like a cheap hmac or signature over the full transaction contents which *must* be the first thing verified. But I think that's a post-mainnet discussion.

Reply:
Reverted skipping the execution check, now only aborting Fee early and added a test for it